File: algorithm/sorting/merge sort/Merge Sort/src/MergeSort.java
public class MergeSort {

    public static int[] mergeSort(int[] inArr) {
        int n = inArr.length;
        if (n <= 1) {
            return inArr;
        }
        int mid = n / 2;
        int[] left = new int[mid];
        int[] right = new int[n - mid];
        for (int i = 0; i < mid; i++) {
            left[i] = inArr[i];
        }
        for (int i = mid; i < n; i++) {
            right[i - mid] = inArr[i];
        }
        int[] sortedLeft = mergeSort(left);
        int[] sortedRight = mergeSort(right);
        return merge(sortedLeft, sortedRight);
    }

    public static int[] merge(int[] left, int[] right) {
        int i = 0, j = 0, k = 0;
        int[] arr = new int[left.length + right.length];
        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }
        while (i < left.length) {
            arr[k++] = left[i++];
        }
        while (j < right.length) {
            arr[k++] = right[j++];
        }
        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {29, 24, 20, 37, 14, 10, 9, 30};
        int[] sortedArr = mergeSort(arr);
        for (int i = 0; i < sortedArr.length; i++) {
            System.out.print(sortedArr[i] + " ");
        }
    }

}

File: Algorithms/Sorting/Merge_Sort/Merge_Sort.py
# Made by Mike_Zhang
# Merge Sort O(n log n)
def merge(left,right):
    result = list()
    while len(left)>0 and len(right)>0:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    while len(left)>0:
        result.append(left.pop(0))
    while len(right)>0:
        result.append(right.pop(0))
    return result

def mergesort(x):
    if len(x)==0 or len(x)==1:
        return x
    else:
        middle = int(len(x)/2)
        a = mergesort(x[:middle])
        b = mergesort(x[middle:])
        return merge(a,b)
    
def main():
    aList=[10,5,2,9,6,3,4,8,1,7]
    print(mergesort(aList))
    list1 = [1,4,6,10]
    list2 = [5,7,8,11,12]
    print(merge(list1,list2))
    
main()
# Made by Mike_Zhang
# Merge Sort O(n log n)